src/extra/JsonSchemaRenderer/index.tsx

```diff
@@ -0,0 +1,213 @@
+import * as React from 'react';
+import { forEach, keys, pick, get, isEqual, isArray } from 'lodash';
+import jsone from 'json-e';
+import ajv from 'ajv';
+import asRendition from '../../asRendition';
+import {
+	DefaultProps,
+	RenditionSystemProps,
+	Theme,
+	Tooltip,
+} from '../../common-types';
+import { Flex } from '../../components/Flex';
+import Alert from '../../components/Alert';
+import widgets, { WidgetWrapperUiOptions, WidgetMeta } from './widgets';
+import { Value, JSONSchema, UiSchema, Format } from './types';
+
+interface WrapperProp extends Tooltip {
+	children: React.ReactNode;
+}
+
+const Wrapper = ({ tooltip, children, ...props }: WrapperProp) => (
+	<Flex
+		tooltip={tooltip}
+		minWidth={0}
+		minHeight={0}
+		alignItems="flex-start"
+		mb={1}
+		flexDirection="column"
+		{...props}
+	>
+		{children}
+	</Flex>
+);
+
+export const getValue = (value?: Value, uiSchema?: UiSchema) => {
+	return get(uiSchema, 'ui:value', value);
+};
+
+export const getType = (value?: Value) => {
+	if (value === undefined) {
+		return 'default';
+	}
+	if (value === null) {
+		return 'null';
+	}
+	return isArray(value) ? 'array' : typeof value;
+};
+
+export const getWidget = (value?: Value, uiSchemaWidget?: string) => {
+	const typeWidgets = get(widgets, getType(value), widgets.default);
+	return get(typeWidgets, uiSchemaWidget || 'default', widgets.default.default);
+};
+
+const processValue = ({
+	value,
+	uiSchema,
+}: Pick<InternalJsonSchemaRendererProps, 'value' | 'uiSchema'>) => {
+	return getValue(value, uiSchema);
+};
+
+class JsonSchemaRenderer extends React.PureComponent<
+	ThemedJsonSchemaRendererProps,
+	JsonSchemaRendererState
+> {
+	validator: ajv.Ajv;
+	validateFn: ajv.ValidateFunction;
+
+	constructor(props: ThemedJsonSchemaRendererProps) {
+		super(props);
+		this.state = {
+			error: null,
+			validationErrors: null,
+		};
+		this.validator = new ajv();
+
+		forEach(props.extraFormats, ({ name, format }) => {
+			this.validator.addFormat(name, format);
+		});
+	}
+
+	componentDidMount() {
+		if (this.props.validate && !this.props.nested) {
+			this.validate(true);
+		}
+	}
+
+	componentDidUpdate(prevProps: ThemedJsonSchemaRendererProps) {
+		if (this.props.validate) {
+			const schemaChanged = !isEqual(prevProps.schema, this.props.schema);
+			if (
+				!this.props.nested &&
+				(!isEqual(prevProps.value, this.props.value) || schemaChanged)
+			) {
+				this.validate(schemaChanged);
+			}
+		}
+	}
+
+	componentDidCatch(error: any) {
+		this.setState({ error });
+	}
+
+	validate(compile: boolean = false) {
```

Comment:
Let's turn this and the related functions into pure ones that take in all necessary parameters (schema, value) and take it out of the component. It can then just return the new state (if we still go with the class approach)

Reply:
Converting to a functional component with hooks has simplified this a lot so I think we're good now!